### ISSUE TITLE
ConcurrentModificationException of the HashMap topicListeners in aws-android-sdk-iot

### ISSUE THREAD

**Issue by claudiopastorini** [aws/aws-sdk-android#329]
I solved this bug, see the attached pull request.

* What service are you using?
**aws-android-sdk-iot**

* In what version of SDK are you facing the problem?
**2.2.22(but with newer version too, the problem persist)** 

* Is the issue limited to Simulators / Actual Devices?
**No**

* Can your problem be resolved if you bump to a higher version of SDK?
**No, but I resolved using a `ConcurrentHashMap`, please see the attached pull request** 

* Is this problem related to specific Android/OS version?
**No**

* Can you give us steps to reproduce with a minimal, complete, and verifiable example? Please include any specific network conditions that might be required to reproduce the problem.
**The problem is related to the concurrent access to the `topicListeners`. I am facing this problem when a thread subscribes a new topic and, in the meanwhile, another thread is informing about the arrival of a new message.**

* Please include a stacktrace if applicable.

```
MqttException (0) - java.util.ConcurrentModificationException
    at org.eclipse.paho.client.mqttv3.internal.CommsCallback.run(CommsCallback.java:197)
    at java.lang.Thread.run(Thread.java:761)
Caused by: java.util.ConcurrentModificationException
    at java.util.LinkedHashMap$LinkedHashIterator.nextEntry(LinkedHashMap.java:411)
    at java.util.LinkedHashMap$KeyIterator.next(LinkedHashMap.java:422)
    at com.########.android.app.AwsClient$2$4.onMessageArrived(AwsClient.java:248)
    at com.amazonaws.mobileconnectors.iot.AWSIotMqttManager$6.messageArrived(AWSIotMqttManager.java:1233)
    at org.eclipse.paho.client.mqttv3.internal.CommsCallback.deliverMessage(CommsCallback.java:475)
    at org.eclipse.paho.client.mqttv3.internal.CommsCallback.handleMessage(CommsCallback.java:379)
    at org.eclipse.paho.client.mqttv3.internal.CommsCallback.run(CommsCallback.java:183)
	... 1 more
```

**Comment by kvasukib** [comment]
Hi @claudiopastorini,

Thanks for reporting to us. We are looking into your pull request.

Meanwhile, can you provide the steps for reproducing the ConcurrentModificationException?

Thanks

**Comment by claudiopastorini** [comment]
Yeah sure.
As I already said, when I try to subscribe a topic in a thread and in the meanwhile the MQTT client receives a message, there is a **ConcurrentModificationException** because of the MQTT client, in order to pass the message to the callback provided with the topic, it iterates over a HashMap that the first thread attempt to modify in order to add a topic and its callback.

1. Create and connect to the MQTT client
2. Subscribe to a topic (eg. topic "A")
3. 
    I. Send a message to the previous token ("A")
    II. In the meanwhile subscribe to another topic (eg. topic "B")

The client disconnects and reconnects after a while (due to the auto-reconnect). In order to obtain the stack trace, I put a breakpoint on line [1241](https://github.com/aws/aws-sdk-android/blob/master/aws-android-sdk-iot/src/main/java/com/amazonaws/mobileconnectors/iot/AWSIotMqttManager.java#L1241) of the [AWSIotMqttManager](https://github.com/aws/aws-sdk-android/blob/master/aws-android-sdk-iot/src/main/java/com/amazonaws/mobileconnectors/iot/AWSIotMqttManager.java) .

**Comment by roshanrajaratnam** [comment]
I can confirm @claudiopastorini issue, I'm facing the same problem. it can be reproduced in the following way as well

1. Create and connect to the MQTT client
2. Subscribe to a topic (eg. topic "A")
3. I. Send a message to the previous token ("A")
    II. Message received from topic A
    III. And subscribe to another topic (eg. topic "B")

**Comment by rohandubal** [comment]
Hello,

This will be fixed with next release of the SDK. I have approved the changes in PR.

Thanks,
Rohan

**Comment by minbi** [comment]
This has been released in version 2.6.19 of the SDK.
